### PR TITLE
Feature/fix zipkin spans alternate

### PR DESCRIPTION
**Pull request by samratdhillon**
Fixing CAMEL-16509 

Cleaner way of implementing copy safe properties on ExtendedExchange

### CODE REVIEW

File: core/camel-api/src/main/java/org/apache/camel/CamelCopySafeProperty.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel;
+
+/**
+ *
+ * @author     Samrat.Dhillon
```
Comment (davsclaus) [inline core/camel-api/src/main/java/org/apache/camel/CamelCopySafeProperty.java:21]: Please remove author tags

File: core/camel-api/src/main/java/org/apache/camel/CamelCopySafeProperty.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * @param  <T>
+ *
+ *             Interface that allows safe copy of property value object when creating copy of Exchange objects. Classes
```
Comment (davsclaus) [inline core/camel-api/src/main/java/org/apache/camel/CamelCopySafeProperty.java:25]: Can you cleanup the javadoc of this

File: core/camel-api/src/main/java/org/apache/camel/CamelCopySafeProperty.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ *             implementing this interface are responsible for creating deep copy of the property value object.
+ */
+public interface CamelCopySafeProperty<T> {
+
+    CamelCopySafeProperty<T> safeCopy();
```
Comment (davsclaus) [inline core/camel-api/src/main/java/org/apache/camel/CamelCopySafeProperty.java:30]: Add javadoc here, as its part of camel-api which we should have documented

File: core/camel-api/src/main/java/org/apache/camel/CamelCopySafeProperty.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ *
+ * @author     Samrat.Dhillon
+ *
+ * @param  <T>
+ *
+ *             Interface that allows safe copy of property value object when creating copy of Exchange objects. Classes
+ *             implementing this interface are responsible for creating deep copy of the property value object.
+ */
+public interface CamelCopySafeProperty<T> {
```
Comment (davsclaus) [inline core/camel-api/src/main/java/org/apache/camel/CamelCopySafeProperty.java:28]: Remove Camel in the name, and maybe `SafeCopyProperty` is a bit better name

File: core/camel-api/src/main/java/org/apache/camel/ExtendedExchange.java [hunk 5]
```diff
@@ -204,4 +204,22 @@
      */
     void setDefaultConsumerCallback(AsyncCallback callback);
 
+    /**
+     * Method to set properties which for which deep copies can be created
```
Comment (davsclaus) [inline core/camel-api/src/main/java/org/apache/camel/ExtendedExchange.java:208]: Polish javadoc, and same for method below

File: core/camel-support/src/main/java/org/apache/camel/support/AbstractExchange.java [hunk 6]
```diff
@@ -886,4 +912,32 @@ public String toString() {
...
+            copySafeProperties.remove(key);
+        }
+
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> T getCopySafeProperty(String key, Class<T> type) {
+
+        Object value = getCopySafeProperties().get(key);
```
Comment (davsclaus) [inline core/camel-support/src/main/java/org/apache/camel/support/AbstractExchange.java:934]: Optimize to check if there is any properties at all (ie its null) to then return null quickly

Comment (samratdhillon) [reply core/camel-support/src/main/java/org/apache/camel/support/AbstractExchange.java]: Thanks @davsclaus . Updated the PR based on your feedback.

**Review by davsclaus — APPROVED** [review]

**Comment by davsclaus** [comment]
Can you work on the other tracer components to let them use safe copy too?

**Comment by samratdhillon** [comment]
> Can you work on the other tracer components to let them use safe copy too?

@davsclaus  I will take a look at opentracing next. However is it okay if I backport this? We are currently working with 3.7.3 and a fixed zipkin implementation will be very helpful.

**Comment by davsclaus** [comment]
ok you can do that

**Pull request merged**